### ISSUE TITLE
Built-in which-key plugin

### ISSUE THREAD

**Issue by brianhuster** [neovim/neovim#31745]
### Problem

Currently, one of the barrier to keyboard-centric editor like Vim/Neovim is that there are too many keymappings to remember. And this is just worse when plugins also introduce their own mappings. 

Both Emacs and Helix has something like [which-key.nvim](https://github.com/folke/which-key.nvim) built-in. I don't know since when has Helix implemented it, but this is [the commit that added the package to Emacs](https://github.com/emacs-mirror/emacs/commit/fa4203300fde6820a017bf1089652fb95759d68c#diff-21241810a84cc186b99d148f58af8e3725198fd1bdad73a9ed5c44bb93b52e88R264). 

### Expected behavior

When users type a key, a floating window (like `which-key.nvim`) or a popup menu (like Helix) open to show users key bindings that start with the key(s) users have typed. 

*** How does it solve the problem?

This reduces the burden of having to remember too many keymaps, while also helping users remember them and use the keyboard more effectively. Also, it will make Neovim friendlier to beginners

An alternative would be just installing a plugin. But many people don't even know such plugins exist, or how they look like. 

*** Is it worth the cost?

I believe so. As long as Neovim and keyboard-centric modal editing exists, the problem would still exists, and the plugin would always be valuable.

**Comment by clason** [comment]
You are not addressing why `:map` ("which-key at home") is not sufficient(!) for the purpose of documentation and discoverability, and why such a (convenience?) feature would _need_ to be in core.

**Comment by brianhuster** [comment]
I think `:map` only works with user-defined or plugin-defined mapping, right? It doesn't work with built-in mapping (defined by C core, I guess). For example, I want to delete a paragraph, but I can't remember the exact keymap for paragraph text object. Both `map da` and `map a` don't work. But if I use `which-key.nvim`, it can show me that the keybindings for `a paragraph` is `ap`. 

I think of `which-key.nvim` like cmdline-completion. Assume that I **can not** use `<Tab>` to toggle cmdline-completion, I can still use `command` to search for the user command (yes, similar to `map`, `command` doesn't work with built-in command), for example `command Co`. The trick works, but it so inconvenient compared to `<Tab>` completion. Not to say many people don't know these tricks. For example, I didn't know that I can use `:map` to look for user keymaps until you told me, I thought it was to create new new keymappings only.

**Comment by Shougo** [comment]
I understand your request. The request itself is very reasonable, and some text editors have it built-in, and other users would probably want it as well.
The issue with this feature is that it is difficult to implement.
Have you ever seen the internal implementation of Vim/Neovim mappings?
Unfortunately, Vim/Neovim mappings are implemented with a large number of `if`/`switch` statements and are not organized. In the first place, there is no "method to get a list of built-in mappings" internally at the moment.
It seems impossible to implement this without completely rewriting the core state transition part. That is very risky.

**Comment by dyfrgi** [comment]
One thing which is implemented separately in both which-key and mini.clue is the naming of keybind prefixes. which-key calls them groups and mini.clue calls them clues, but in both cases they consist of a set of modes, an input prefix, and a name. I can't find a ticket for it specifically, but this would also be useful for a command palette as it allows showing just the parent node of a tree of commands rather than the entire list of commands.

Having a built-in concept of command prefixes seems likely to be necessary to enable an ecosystem of "out-of-the-box" installable plugins that serve this need. I think this is especially important for capturing bindings created by other plugins which otherwise require additional manual configuration. Sometimes people use bindings to <Nop> to label groups, but I think that's probably going to cause some weirdness with input timeouts given https://github.com/neovim/neovim/issues/18877.

I think that adding such groups would be a requirement for this feature, but could also be done without integrating some specific script that also handles things like showing options based on the current input state and command search.

**Comment by justinmk** [comment]
> naming of keybind prefixes. which-key calls them groups and mini.clue calls them clues, but in both cases they consist of a set of modes, an input prefix, and a name. 

sounds like maybe "namespaces" would cover that. 

- https://github.com/neovim/neovim/issues/32261#issuecomment-2635417146
- https://github.com/neovim/neovim/issues/34704

**Comment by dyfrgi** [comment]
namespaces are related but I think that they're nearly orthogonal. My understanding of namespaces is that it's about enabling multiple plugins to map the same key and having behavior besides "last map applied wins" or "error because you didn't say you were remapping". Prefix groups is about giving a name to an intermediate node in the nominal state machine (I don't think neovim actually models it this way, but conceptually) that decides what to do with some user input.

e.g. suppose I have three bindings to `<leader>ax`, `<leader>ay`, and `<leader>az`. When I press `<leader>`, a plugin like which-key or mini.clue has to show something for what happens when I press `a`. The group or clue gives that intermediate state you're in after `<leader>a` a name which can be shown in that hint instead of just saying that there's 3 bindings.

I think that the overlap between the two feature happens when you have conflicts between prefixes. If I install two plugins with default bindings and both put them behind `<leader>a`, the namespace to pull commands from could be selected at that point rather than when I finish. And without doing that any hints will be confusing, especially if the design you sketched where Nvim opens up a menu of namespaces after you finish typing the command is selected.

**Comment by justinmk** [comment]
> e.g. suppose I have three bindings to `<leader>ax`, `<leader>ay`, and `<leader>az`. When I press `<leader>`, a plugin like which-key or mini.clue has to show something for what happens when I press `a`. The group or clue gives that intermediate state you're in

yeah but that's a workaround for the lack of namespaces or some other "key space". a which-key interface removes the need for manually arranging shared prefixes, by design. or maybe i'm thinking of a fuzzy picker.
